aten/src/ATen/native/native_functions.yaml

```diff
@@ -109,6 +109,12 @@
 
 - func: matmul(Tensor self, Tensor other) -> Tensor
 
+# we define both of these because 'where' does the broadcast and '_s_where' doesn't;
+# this allows us to implicitly calculate the broadcast derivative, while only dealing with the
+# _s_where derivative.
+- func: where(BoolTensor condition, Tensor self, Tensor other) -> Tensor
+- func: _s_where(BoolTensor condition, Tensor self, Tensor other) -> Tensor
```

Comment:
I'm not sure if I understand the comment. The downside of doing it like this is that we'll have `_s_where` show up in JIT traces.

Reply:
The comment is just related to how we usually specify derivatives of functions that broadcast; i.e. the broadcast is handled separately.  Since we don't have all the mechanism to do that in native functions (and it's not clear we want it), we fake it.

I agree on the downside.  @ezyang have you thought about this?

Reply:
`_s_where` won't show up, because currently (after trace native functions) only trace the outermost native function. With scopes you'll also get to see the unfolding. I don't think there is any downside for tracing.

More generally, it seems the problem is that we're currently automatically generating the broadcast/non-broadcast versions of code, but in a way that is not easy to make use of from native. It seems to me that broadcasting is common enough that there should be an affordance for it, also for native.

Reply:
Cool, thanks @ezyang.

On the broadcasting code, I was hoping to avoid that because although the common case of broadcasting is simple, the general case is complex (e.g. torch.matmul has rules that are totally different from everything else).  If you just write what you want to happen in native_functions and it works that seems like the ideal solution to me.

Reply:
I still don't fully understand why can't we use `where` for the derivative - yes, it's broadcasting and it's not needed in the backwards, but it would still be correct, right?

@ezyang  I don't know how is it going to get DCEd, if you call `_s_where`, there's no `where` outside to take over the trace.

Reply:
Who is you?  Users should always call `where`.

Reply:
you is the code for the derivative. I assume it will look approximately like this: `_s_where(cond, grad_x, grad_y)`, so there's no outer `where` to capture the trace.

Reply:
This also shows us that we're doing some unnecessary work by allocating zeros twice

Reply:
Oh, I see what you mean now about _s_where in the backwards; I thought you were talking about using `func: where` (instead of` func: _s_where`) , not using `where` in the definition of the backward.  We can use `where` in the definition of the backward, I did the above to avoid the broadcasting checks, but we don't do that in general.

Reply:
Yes, that's true. But the non-broadcast version is simply a special case of the broadcasting version, so it shouldn't be hard for the JIT to handle both, if it needs to handle them. No consequence for ONNX which doesn't export backward (yet).